### PR TITLE
Rustc explain

### LINKED ISSUE #48041: Put back "try with rustc --explain E0XXX"
This sentence has been removed [here](https://github.com/rust-lang/rust/commit/fad4f32c3#diff-4d5b8cff35812b800ff77376382671a7L781) in [this PR](https://github.com/rust-lang/rust/pull/35401). However, I think it's quite useful considering the amount of extra information it provides.

cc @rust-lang/compiler

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #48041.

To make the review easier, I separated tests update to code update. Also, I used this script to generate new ui tests stderr:

```python
from os import listdir
from os.path import isdir, isfile, join

PATH = "src/test/ui"

def do_something(path):
    files = [join(path, f) for f in listdir(path)]

    for f in files:
        if isdir(f):
            do_something(f)
            continue
        if not isfile(f) or not f.endswith(".stderr"):
            continue
        x = open(f, "r")
        content = x.read().strip()
        if "error[E" not in content:
            continue
        errors = dict()
        for y in content.splitlines():
            if y.startswith("error[E"):
                errors[y[6:11]] = True
        errors = sorted(errors.keys())
        if len(errors) < 1:
            print("weird... {}".format(f))
            continue
        if len(errors) > 1:
            content += "\n\nYou've got a few errors: {}".format(", ".join(errors))
            content += "\nIf you want more information on an error, try using \"rustc --explain {}\"".format(errors[0])
        else:
            content += "\n\nIf you want more information on this error, try using \"rustc --explain {}\"".format(errors[0])
        content += "\n"
        x = open(f, "w")
        x.write(content)

do_something(PATH)
```

### CODE REVIEW

File: src/librustc_errors/emitter.rs [hunk 1]
```diff
@@ -115,6 +116,30 @@ struct FileWithAnnotatedLines {
...
+            if error_codes.len() > 1 {
+                writeln!(self.dst,
+                         "You've got a few errors: {}",
+                         error_codes.join(", ")).expect("failed to give tips...");
+                writeln!(self.dst,
+                         "If you want more information on an error, try using \
+                          \"rustc --explain {}\"",
+                         &error_codes[0]).expect("failed to give tips...");
+            } else {
+                writeln!(self.dst,
```
Comment (oli-obk) [inline src/librustc_errors/emitter.rs]: Should this message show up for errors that don't show any extra information if `--explain` is run?

**Comment by oli-obk** [comment]
> Did I miss something (it's just like the code is run in parts and then the new error message gets printed after each of them)?

That might be because ui tests are run with `--error-format=json" and the `rendered` field of all diagnostics are concatenated.

**Comment by GuillaumeGomez** [comment]
Should I update how it's done then? Because right now, as I said, it has nothing in common with the actual output...

**Comment by GuillaumeGomez** [comment]
Should be done now.

**Comment by estebank** [comment]
Still failing :-/

[Minimal difference due to JSON encoding](https://travis-ci.org/rust-lang/rust/builds/343589013#L1597):

```
[01:01:33] ---- [ui] ui/lint/use_suggestion_json.rs stdout ----
[01:01:33] 	diff of stderr:
[01:01:33] 
[01:01:33] 384	   |
[01:01:33] 385	and 8 other candidates
[01:01:33] 386	
[01:01:33] -	If you want more information on this error, try using "rustc --explain E0412"
[01:01:33] +	If you want more information on this error, try using /"rustc --explain E0412/"
[01:01:33] 388	"
[01:01:33] 389	}
[01:01:33] 390	{
```

File: src/tools/compiletest/src/runtest.rs [hunk 2]
```diff
@@ -1629,8 +1632,8 @@ impl<'test> TestCx<'test> {
                 .iter()
                 .any(|s| s.starts_with("--error-format"))
             {
-                rustc.args(&["--error-format", "json"]);
-            },
+                //rustc.args(&["--error-format", "json"]);
```
Comment (oli-obk) [inline src/tools/compiletest/src/runtest.rs]: leftover comment.

**Review by oli-obk — COMMENTED** [review]
The windows/linux normalizer changes `\` to `/`, previously the json decoder removed escape backslashes during string decoding

**Comment by GuillaumeGomez** [comment]
It passed tests. \o/

File: src/librustc_errors/emitter.rs [hunk 3]
```diff
@@ -115,6 +116,30 @@ struct FileWithAnnotatedLines {
     multiline_depth: usize,
 }
 
+impl Drop for EmitterWriter {
+    fn drop(&mut self) {
+        if !self.short_message && !self.error_codes.is_empty() {
+            let mut error_codes = self.error_codes.clone().into_iter().collect::<Vec<_>>();
+            error_codes.sort();
+            if error_codes.len() > 1 {
```
Comment (estebank) [inline src/librustc_errors/emitter.rs]: Should this restrict the error list to some reasonable amount? I'm ok with any amount, but I'd constrain it to at most two full 80 col lines.

**Review by estebank — APPROVED** [review]
After addressing the nitpick, r=me.

File: src/librustc_errors/emitter.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustc_errors/emitter.rs]: Does it really matter? I can do it but I'm not sure if this is very useful... You're not supposed to try to have as much errors as possible when writing code haha.

Comment (estebank) [reply src/librustc_errors/emitter.rs]: It should be rare, but I see no reason for us not to gracefully handle that case :)

It won't ever be possible for a given project to have more than a subset of the ~600 errors the compiler has, but regardless, it is worthless to list the error codes, once the user understands how to use the flag, it makes more sense to scroll to the error that they want more information on and copy the error code from there, as the error code on it's own is meaningless.

**Comment by estebank** [comment]
I still think there's value in limiting the amount of error codes listed to a reasonable amount, but it is silly to block on it.

Also, it would also be nice to have a way to disable this output, but I won't block this PR on this. We need to add a compiler config file to handle this and other cases (--teach, coloring, one line output, etc.).

@bors r+

File: src/librustc_errors/emitter.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustc_errors/emitter.rs]: Ok, I'll limit then.

**Comment by GuillaumeGomez** [comment]
Updated and limited error codes list output.

**Comment by estebank** [comment]
https://travis-ci.org/rust-lang/rust/builds/345893093#L1616

```
[00:53:13] failures:
[00:53:13]     [ui] ui/feature-gate-if_while_or_patterns.rs
[00:53:13]     [ui] ui/impl_trait_projections.rs
[00:53:13]     [ui] ui/issue-23302-1.rs
[00:53:13]     [ui] ui/issue-23302-2.rs
[00:53:13]     [ui] ui/issue-23302-3.rs
[00:53:13]     [ui] ui/issue-36163.rs
[00:53:13]     [ui] ui/issue-45157.rs
[00:53:13]     [ui] ui/issue-47706.rs
[00:53:13]     [ui] ui/nested_impl_trait.rs
```

**Comment by GuillaumeGomez** [comment]
Damn, it was new tests...

**Comment by GuillaumeGomez** [comment]
CI passed so let's r+ again (with a higher priority to avoid being outdated once again)...

@bors: r=estebank p=10

File: src/tools/compiletest/src/runtest.rs [hunk 4]
```diff
@@ -2544,6 +2545,7 @@ impl<'test> TestCx<'test> {
             }
         }
         if !explicit {
+            let proc_res = self.compile_test(&["--error-format", "json"]);
```
Comment (petrochenkov) [inline src/tools/compiletest/src/runtest.rs]: Yeah, let's just compile all the tests twice.
Who cares about time, certainly [not Appveyor](https://github.com/rust-lang/rust/issues/46903#issuecomment-357061260).

**Comment by zackmdavis** [comment]
@GuillaumeGomez 

> If you want more information on an error, try using "rustc --explain {}"

I know I'm late, but I would argue that this would be better phrased as something more like:
```
For more information about an error, try `rustc --explain {}`
```
(avoids informal-sounding second person, omits needless word 'using', backticks instead of quotation marks for code).

**Comment by GuillaumeGomez** [comment]
@zackmdavis: A bit yes. Let me update then...

File: src/librustc_errors/emitter.rs [hunk 5]
```diff
@@ -115,6 +116,33 @@ struct FileWithAnnotatedLines {
     multiline_depth: usize,
 }
 
+impl Drop for EmitterWriter {
+    fn drop(&mut self) {
```
Comment (zackmdavis) [inline src/librustc_errors/emitter.rs]: @GuillaumeGomez (more belated commentary in light of #48550) What was the motivation for putting the this in `drop()`? Semantically, printing the message notifying the user about `--explain` does not seem like a "cleanup" action. As [Niko suggested](https://github.com/rust-lang/rust/issues/48041#issuecomment-363842863), I would kind of expect this to go with [the "aborting due to previous error" message in `Handler.abort_if_errors`](https://github.com/rust-lang/rust/blob/89e5a0796e0c5de52cd1d6c7541e5c2c7c47f2cd/src/librustc_errors/lib.rs#L562-L564)—and it looks like [the `Handler` already knows which codes have been emitted](https://github.com/rust-lang/rust/blob/89e5a0796e0c5de52cd1d6c7541e5c2c7c47f2cd/src/librustc_errors/lib.rs#L604-L610), too.

Comment (GuillaumeGomez) [reply src/librustc_errors/emitter.rs]: Hum, if this is true, it might allow to make my task for #48562.

**Comment by zackmdavis** [comment]
@GuillaumeGomez One _last_ comment (sorry if it seems like I've been overly critical with respect to this feature; sometimes good code requires a lot of discussion, and sometimes I have a lot of things to say), but arguably the `--explain` usage message should come _before_ the "aborting due to _n_ errors" message (so that we're aborting immediately after we say that we're aborting).

**Comment by matklad** [comment]
TIL that you can not only `rustc --explain CODE`, but `cargo --explain CODE` as well: https://github.com/rust-lang/cargo/pull/2551

The commit message on that PR implies the idea was to make `rustc` print just `run --explain CODE` and not `run rustc --explain CODE` or something like this (I am not sure I understand this entirely).

**Pull request merged**